python/cudf/cudf/core/index.py

```diff
@@ -1411,24 +1488,527 @@ def __cuda_array_interface__(self):
+        out = Frame().__new__(cls)
+
+        out._initialize_numeric(
+            data=data,
+            dtype=dtype,
+            expected_dtype="float64",
+            copy=copy,
+            name=name,
+        )
+
+        return out
+
+
 class DatetimeIndex(GenericIndex):
-    # TODO this constructor should take a timezone or something to be
-    # consistent with pandas
-    def __new__(cls, values, **kwargs):
+    """
+    Immutable ndarray of datetime64 data, represented internally as int64.
+
+    Parameters
+    ----------
+    data : array-like (1-dimensional), optional
+        Optional datetime-like data to construct index with.
+    copy : bool
+        Make a copy of input ndarray.
+    freq : str, optional
+        This is not yet supported
+    tz : pytz.timezone or dateutil.tz.tzfile
+        This is not yet supported
+    ambiguous : ‘infer’, bool-ndarray, ‘NaT’, default ‘raise’
+        This is not yet supported
+    name : object
+        Name to be stored in the index.
+    dayfirst : bool, default False
+        If True, parse dates in data with the day first order.
+        This is not yet supported
+    yearfirst : bool, default False
+        If True parse dates in data with the year first order.
+        This is not yet supported
+
+    Returns:
+    --------
+    DatetimeIndex
+
+    Examples
+    --------
+    >>> import cudf
+    >>> cudf.DatetimeIndex([1, 2, 3, 4], name="a")
+    DatetimeIndex(['1970-01-01 00:00:00.001000', '1970-01-01 00:00:00.002000',
+                   '1970-01-01 00:00:00.003000', '1970-01-01 00:00:00.004000'],
+                  dtype='datetime64[ms]', name='a')
+    """
+
+    def __new__(
+        cls,
+        data=None,
+        freq=None,
+        tz=None,
+        normalize=False,
+        closed=None,
+        ambiguous="raise",
+        dayfirst=False,
+        yearfirst=False,
+        dtype=None,
+        copy=False,
+        name=None,
+    ) -> "DatetimeIndex":
         # we should be more strict on what we accept here but
         # we'd have to go and figure out all the semantics around
         # pandas dtindex creation first which.  For now
         # just make sure we handle np.datetime64 arrays
         # and then just dispatch upstream
         out = Frame().__new__(cls)
-        kwargs = _setdefault_name(values, kwargs)
-        if isinstance(values, np.ndarray) and values.dtype.kind == "M":
-            values = column.as_column(values)
-        elif isinstance(values, pd.DatetimeIndex):
-            values = column.as_column(values.values)
-        elif isinstance(values, (list, tuple)):
-            values = column.as_column(np.array(values, dtype="<M8[ms]"))
-        out.__initializer__(values, **kwargs)
+
+        if freq is not None:
+            raise NotImplementedError("Freq is not yet supported")
+        if tz is not None:
+            raise NotImplementedError("tz is not yet supported")
+        if normalize is not False:
```

Comment:
If someone explicitly passes `0` then this would raise erroneously:
```suggestion
        if normalize != False:
```